packages/bootstrap-vue-next/src/components/BForm/BFormDatalist.vue

```
@@ -0,0 +1,56 @@
+<template>
+  <datalist :id="computedId">
+    <slot name="first" />
+    <template v-for="(option, index) in normalizedOptsWrapper" :key="index">
+      <BFormSelectOption :value="option.value" :disabled="option.disabled">
+        <!-- eslint-disable-next-line vue/no-v-html -->
+        <span v-if="!!option.html" v-html="option.html" />
+        <template v-else>
+          {{ option.text }}
+        </template>
+      </BFormSelectOption>
+    </template>
+    <slot />
+  </datalist>
+</template>
+
+<script setup lang="ts" generic="T">
+import type {BFormSelectProps, SelectOption} from '../../types'
+import {computed} from 'vue'
+import BFormSelectOption from '../BFormSelect/BFormSelectOption.vue'
+import {useDefaults, useFormSelect, useId} from '../../composables'
+
+const _props = withDefaults(defineProps<BFormSelectProps>(), {
+  disabled: false,
+  disabledField: 'disabled',
+  htmlField: 'html',
+  id: undefined,
+  // eslint-disable-next-line vue/require-valid-default-prop
+  options: () => [],
+  textField: 'text',
+  valueField: 'value',
+})
+const props = useDefaults(_props, 'BFormSelect')
+
+defineSlots<{
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  default?: (props: Record<string, never>) => any
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  first?: (props: Record<string, never>) => any
+}>()
+
+const computedId = useId(() => props.id, 'datalist')
+
+const {normalizedOptions, isComplex} = useFormSelect(() => props.options, props)
+
+const normalizedOptsWrapper = computed(() => {
+  let opts = normalizedOptions.value
+  if (opts.find((opt) => isComplex(opt))) {
+    // eslint-disable-next-line no-console
+    console.warn('Datalist may not contain complex options', computedId.value)
```

I'd like to use the same mechansisms we have in place for options lists, but datafields don't allow for options groups. I'm not seeing a consistent way to manage runtime errors for a case like this, happy with any guidance for managing this differently.

---

You'd have to show some code for what you tried.

I don't like this code here. 

I think if this component never has a situation where isComplex should be true, then it should never be an issue and this code should be removed. Ie if it does ever happen somehow, we should simply ignore it as its user error

Hypothetically this is something that would be validated by TS. Don't know how strongly typed that is though

---

Here's a fiddle https://jsfiddle.net/ohdwg/6tzu4nqc/2/ showing that options groups aren't supported in data lists.  They're also not documented on Mozilla for datalist: https://jsfiddle.net/ohdwg/6tzu4nqc/2/ while they are for  select: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select

I agree that typing the property would be the preferred way to go. But, the types have to be loose enough to allow substituting key names for the different fields (diabledField, htmlField, textField, valueField, labelField, optionsField).

I reworked this to silently ignore complex options and addressed the limitation in the documentation. I considered flattening the list (which is closer to what the native control does if you add in complex options), but then I'd have to add back in `labelField` and `optionsField` which seems like ugly complexity for something that is getting flattened back out. Short of pulling off some typescript voodoo, that's the best I can come up with.